rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,296 @@
+
+#### Privacy
+
+Events will _only_ be collected and sent under the following conditions:
+
+- An `ANALYTICS_ENABLED` feature flag is set to `true`.
+- The user has not opted out of analytics by setting the `allowAnalytics` value in our UI cookie store to `false`.
+
+Unless all conditions are met the `sendEvent` function will simply no-op.
+
+A previous version of this proposal suggested honoring the `navigator.doNotTrack` browser method, but upon inspection this was deprecated in 2018 and is no longer advised.
+
+This proposal will require the creation of a user settings page _or_ an analytics opt-out checkbox on the `/privacy` page.
+
+### Sample implementaton and event types
+
+Below I've created a TypeScript type for all the proposed anlytics events, and a simple function for sending events. For each event you'll see the payload alongside the default values which will be passed to _all_ events.
+
+```ts
+// src/utils/events.ts
+
+// This type is only for illustrative purposes and would be imported from elsewhere.
+type MediaType = "image" | "audio" | "all";
+
+/**
+ * A list of all events allowed by our analytics server.
+ */
+type AnalyticsEvents = {
+  /** A search performed by the user **/
+  SUBMIT_SEARCH: {
+    /** The media type being searched */
+    mediaType: MediaType;
+    /** The search term **/
+    query: string;
+  };
+  /** The user navigates to a page. */
+  VIEW_PAGE: {
+    /** The title of the page being navigated to (window.document.title)  */
+    title: string;
+    /** The path of the previous page */
+    previous: string;
+    /** The path of the new page */
+    next: string;
+  };
+  /** Whenever the user sets a filter  */
+  APPLY_FILTER: {
+    /** The name of the filter (not the user-facing element label) */
+    key: string;
+    /** The value of the filter */
+    value: string | boolean;
+    /** The media type being searched */
+    mediaType: MediaType;
+    /** The search term **/
+    query: string;
+  };
+  /** Whenever a user changes the content type */
+  CHANGE_CONTENT_TYPE: {
+    /** The previously-set media type */
+    previous: MediaType;
+    /** The new media type */
+    next: MediaType;
+    /** The name of the Vue component used to switch content types. **/
+    component: string;
+  };
+  /**
+   * Whenever the user scrolls to the end of the results page.
+   * Useful to evaluate how often users load more results or click
+   * on the external sources dropdown.
+   * */
+  REACH_RESULT_END: {
+    /** The media type being searched */
+    mediaType: MediaType;
+    /** The search term **/
+    query: string;
+  };
```

Comment:
Similar to `REACH_RESULT_END`, we should record search queries that generate a very small number (`INSUFFICIENT_RESULTS`) or no results at all (`ZERO_RESULTS`).

Reply:
This is a good idea. I wonder if it should be the same event `INSUFFICIENT_RESULTS` but include the result count in the payload. This could be really helpful for making determinations about what types of content users _want_ that we don't currently offer.

I think loading results _less than_ one full page would be a good trigger to the event.